app_legacy/Helpers/database/player-game.php

```diff
@@ -91,60 +91,193 @@ function getGameRankAndScore(int $gameID, string $username): array
-              WHERE ach.GameID IN ( $gameIDs ) AND ach.Flags = " . AchievementType::OfficialCore . "
-              AND aw.User = :username
-              GROUP BY aw.HardcoreMode, ach.GameID";
-
-    $dbResult = legacyDbFetchAll($query, ['username' => $user]);
-    foreach ($dbResult as $data) {
-        if ($data['HardcoreMode'] == 0) {
-            $dataOut[$data['GameID']]['NumAchieved'] = $data['AchCount'];
-            $dataOut[$data['GameID']]['ScoreAchieved'] = $data['PointCount'];
-        } else {
-            $dataOut[$data['GameID']]['NumAchievedHardcore'] = $data['AchCount'];
-            $dataOut[$data['GameID']]['ScoreAchievedHardcore'] = $data['PointCount'];
+    if ($numRecentAchievements >= 0) {
+        usort($unlockedAchievements, function ($a, $b) {
+            if ($a['When'] == $b['When'])
+                return $a['Achievement']['ID'] <=> $b['Achievement']['ID'];
+
+            return -($a['When'] <=> $b['When']);
+        });
+
+        if ($numRecentAchievements !== 0) {
+            $unlockedAchievements = array_slice($unlockedAchievements, 0, $numRecentAchievements);
+        }
+
+        $recentAchievements = [];
+
+        foreach ($unlockedAchievements as $unlockedAchievement) {
+            $gameData = $unlockedAchievement['Game'];
+            $gameID = (int) $gameData['ID'];
+            $achievementData = $unlockedAchievement['Achievement'];
+            $achievementID = (int) $achievementData['ID'];
+
+            $recentAchievements[$gameID][$achievementID] = [
+                'ID' => $achievementID,
+                'GameID' => $gameID,
+                'GameTitle' => $gameData['Title'],
+                'Title' => $achievementData['Title'],
+                'Description' => $achievementData['Description'],
+                'Points' => (int) $achievementData['Points'],
+                'BadgeName' => $achievementData['BadgeName'],
+                'IsAwarded' => '1',
+                'DateAwarded' => $unlockedAchievement['When'],
+                'HardcoreAchieved' => (int) $unlockedAchievement['Hardcore'],
+            ];
         }
+
+        foreach ($lockedAchievements as $lockedAchievement) {
+            $gameData = $lockedAchievement['Game'];
+            $gameID = (int) $gameData['ID'];
+            $achievementData = $lockedAchievement['Achievement'];
+            $achievementID = (int) $achievementData['ID'];
+
+            $recentAchievements[$gameID][$achievementID] = [
+                'ID' => $achievementID,
+                'GameID' => $gameID,
+                'GameTitle' => $gameData['Title'],
+                'Title' => $achievementData['Title'],
+                'Description' => $achievementData['Description'],
+                'Points' => (int) $achievementData['Points'],
+                'BadgeName' => $achievementData['BadgeName'],
+                'IsAwarded' => '0',
+                'DateAwarded' => null,
+                'HardcoreAchieved' => null,
+            ];
+        }
```

Comment:
Appending the locked items mimics the old behavior of `ORDER BY IsAwarded`, and is required to display the still locked items on the user page.